File: Leetcode 75/Day 5 - Greedy/2. Longest Palindrome (409)/LongestPalindrome.java
class Solution {
    public int longestPalindrome(String s) {
        // palindrome: word that is the same backwards and forwards
        // hash map to keep track of values
        // if values are even, add it to the total
        // if values are false, subtract one
        // if there is an odd value, add 1 at the end -> put the odd value in the middle

        Map<Character, Integer> hashMap = new HashMap();
        int longestPalindrome = 0;
        boolean oddFound = false;

        // get each char and the counts into the hash map
        for (Character c : s.toCharArray()) {
            hashMap.put(c, hashMap.getOrDefault(c, 0) + 1);
        }

        for (int value : hashMap.values()) {
            if (value % 2 == 0) {
                longestPalindrome += value;
            } else {
                oddFound = true;
                longestPalindrome += value - 1;
            }
        }

        if (oddFound) {
            longestPalindrome++;
        }
        return longestPalindrome;
    }
}

File: Leetcode 75/Day 5 - Greedy/2. Longest Palindrome (409)/LongestPalindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> int:

        largestPalindrome = 0
        hashMap = {}
        thereIsOdd = False

        for letter in s:
            hashMap[letter] = hashMap.get(letter, 0) + 1

        for v in hashMap.values():
            if v % 2 == 0:
                largestPalindrome += v
            else:
                largestPalindrome += (v-1)
                thereIsOdd = True

        if thereIsOdd:
            return largestPalindrome + 1

        return largestPalindrome
